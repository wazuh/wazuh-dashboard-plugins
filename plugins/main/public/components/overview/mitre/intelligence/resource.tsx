/*
 * Wazuh app - React component for showing the Mitre Att&ck resource items.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */

import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { TableWzAPI } from '../../../../components/common/tables';
import { WzRequest } from '../../../../react-services';
import { ModuleMitreAttackIntelligenceFlyout } from './resource_detail_flyout';
import { UI_LOGGER_LEVELS } from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';
import NavigationService from '../../../../react-services/navigation-service';
import { Route, Switch } from '../../../router-search';

export const ModuleMitreAttackIntelligenceResource = ({
  label,
  searchBar,
  apiEndpoint,
  tableColumnsCreator,
  initialSortingField,
  resourceFilters,
}) => {
  const [details, setDetails] = useState(null);
  const navigationService = NavigationService.getInstance();

  useEffect(() => {
    const urlParams = navigationService.getParams();
    const hasRedirectTabParam = urlParams.has('tabRedirect');
    const hasIdToRedirectParam = urlParams.has('idToRedirect');
    if (hasRedirectTabParam && hasIdToRedirectParam) {
      const redirectTab = urlParams.get('tabRedirect');
      const idToRedirect = urlParams.get('idToRedirect');
      const endpoint = `/mitre/${redirectTab}?q=external_id=${idToRedirect}`;
      getMitreItemToRedirect(endpoint);
    }
  }, [
    navigationService.getParams().has('tabRedirect'),
    navigationService.getParams().has('idToRedirect'),
  ]);

  const getMitreItemToRedirect = async endpoint => {
    try {
      const res = await WzRequest.apiReq('GET', endpoint, {});
      const data = res?.data?.data.affected_items;
      setDetails(data[0]);
    } catch (error) {
      const options = {
        context: `${ModuleMitreAttackIntelligenceResource.name}.getMitreItemToRedirect`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        display: true,
        error: {
          error: error,
          message: error.message || error,
          title: error.name || error,
        },
      };
      getErrorOrchestrator().handleError(options);
    }
  };

  const tableColumns = useMemo(() => tableColumnsCreator(), []);

  const closeFlyout = useCallback(() => {
    setDetails(null);
    NavigationService.getInstance().updateAndNavigateSearchParams({
      idToRedirect: null,
    });
  }, []);

  return (
    <>
      <TableWzAPI
        searchTable
        title={label}
        tableColumns={tableColumns}
        tableInitialSortingField={initialSortingField}
        endpoint={apiEndpoint}
        tablePageSizeOptions={[10, 15, 25, 50, 100]}
        filters={resourceFilters}
        searchBarWQL={{
          options: searchBar.wql.options,
          suggestions: searchBar.wql.suggestions,
        }}
      />
      {details && (
        <Switch>
          <Route
            path='?tabRedirect=:tabRedirect&idToRedirect=:idToRedirect'
            render={() => (
              <ModuleMitreAttackIntelligenceFlyout
                details={details}
                closeFlyout={() => closeFlyout()}
                onSelectResource={setDetails}
              />
            )}
          />
        </Switch>
      )}
    </>
  );
};
