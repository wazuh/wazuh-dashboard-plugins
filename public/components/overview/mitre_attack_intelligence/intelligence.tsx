/*
 * Wazuh app - React component for showing the Mitre Att&ck intelligence.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */

import React, { useCallback, useState, useRef, useEffect } from 'react';
import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';

import { MitreAttackResources } from './resources';
import { ModuleMitreAttackIntelligenceLeftPanel } from './intelligence_left_panel';
import { ModuleMitreAttackIntelligenceRightPanel } from './intelligence_right_panel';
import { useAsyncAction } from '../../common/hooks';
import { WzRequest } from '../../../react-services';
import { PanelSplit } from '../../common/panels';
import { withUserAuthorizationPrompt } from '../../common/hocs';
import { compose } from 'redux';

export const ModuleMitreAttackIntelligence = compose(
  withUserAuthorizationPrompt([{ action: 'mitre:read', resource: '*:*:*' }])
)(() => {
  const [selectedResource, setSelectedResource] = useState(MitreAttackResources[0].id);
  const [searchTermAllResources, setSearchTermAllResources] = useState('');
  const searchTermAllResourcesLastSearch = useRef('');
  const [resourceFilters, setResourceFilters] = useState([]);
  const searchTermAllResourcesUsed = useRef(false);
  const searchTermAllResourcesAction = useAsyncAction(
    async (searchTerm) => {
      selectedResource !== null && setSelectedResource(null);
      searchTermAllResourcesUsed.current = true;
      searchTermAllResourcesLastSearch.current = searchTerm;
      const limitResults = 5;
      return (
        await Promise.all(
          MitreAttackResources.map(async (resource) => {
            const response = await WzRequest.apiReq('GET', resource.apiEndpoint, {
              params: { search: searchTerm, limit: limitResults },
            });
            return {
              id: resource.id,
              name: resource.label,
              fieldName: resource.fieldName,
              results: response?.data?.data?.affected_items,
              totalResults: response?.data?.data?.total_affected_items,
              loadMoreResults:
                response?.data?.data?.total_affected_items &&
                response?.data?.data?.total_affected_items > limitResults &&
                (() => {
                  setResourceFilters([
                    { field: 'search', value: searchTermAllResourcesLastSearch.current },
                  ]);
                  setSelectedResource(resource.id);
                }),
            };
          })
        )
      ).filter((searchTermAllResourcesResponse) => searchTermAllResourcesResponse.results.length);
    },
    [searchTermAllResources]
  );

  useEffect(() => {
    const urlParams = new URLSearchParams(location.href);
    const redirectTab = urlParams.get('tabRedirect');
    if (redirectTab) {
      setSelectedResource(redirectTab);
    }
  }, []);

  const onSelectResource = useCallback(
    (resourceID) => {
      setResourceFilters([]);
      setSelectedResource((prevSelectedResource) =>
        prevSelectedResource === resourceID && searchTermAllResourcesUsed.current
          ? null
          : resourceID
      );
    },
    [searchTermAllResourcesUsed.current]
  );

  const onSearchTermAllResourcesChange = useCallback((searchTerm) => {
    setSearchTermAllResources(searchTerm);
  }, []);

  return (
    <EuiFlexGroup style={{ margin: '0 8px' }}>
      <EuiFlexItem>
        <PanelSplit
          side={
            <ModuleMitreAttackIntelligenceLeftPanel
              onSearchTermAllResourcesChange={onSearchTermAllResourcesChange}
              onSearchTermAllResourcesSearch={searchTermAllResourcesAction.run}
              onSelectResource={onSelectResource}
              selectedResource={selectedResource}
            />
          }
          sideProps={{ style: { width: '15%', minWidth: 145, overflowX: 'hidden' } }}
          content={
            <ModuleMitreAttackIntelligenceRightPanel
              results={searchTermAllResourcesAction.data}
              loading={searchTermAllResourcesAction.running}
              selectedResource={selectedResource}
              resourceFilters={resourceFilters}
            />
          }
          contentProps={{
            style: { maxHeight: 'calc(100vh - 255px)', overflowY: 'auto', overflowX: 'hidden' },
          }}
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  );
});
