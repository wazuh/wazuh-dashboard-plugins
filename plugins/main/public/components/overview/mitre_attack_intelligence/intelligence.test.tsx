/*
 * Wazuh app - ModuleMitreAttackIntelligence Component - Test
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import React from 'react';
import { render } from 'enzyme';
import { ModuleMitreAttackIntelligence } from './intelligence';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

jest.mock(
  '../../../../../../node_modules/@elastic/eui/lib/services/accessibility/html_id_generator',
  () => ({
    htmlIdGenerator: () => () => 'htmlId',
  })
);

jest.mock('../../../react-services', () => ({
  WzRequest: () => ({
    apiReq: (method: string, path: string, params: any) => {
      return {
        data: {
          data: {
            affected_items: [],
          },
        },
      };
    },
  }),
}));

// added to remove useLayoutEffect warning
jest.mock('react', () => ({
  ...(jest.requireActual('react') as object),
  useLayoutEffect: jest.requireActual('react').useEffect,
}));

const mockStore = configureMockStore();

describe('Module Mitre Att&ck intelligence container', () => {
  it('should render the component if has permissions', () => {
    const store = mockStore({
      appStateReducers: {
        withUserLogged: true,
        userRoles: ['administrator'],
        userPermissions: {
          'mitre:read': { '*:*:*': 'allow' },
        },
      },
    });
    const component = render(
      <Provider store={store}>
        <ModuleMitreAttackIntelligence />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('should render permissions prompt when no has permissions', () => {
    const store = mockStore({
      appStateReducers: {
        withUserLogged: true,
        userRoles: ['administrator'],
        userPermissions: {
          'mitre:read': { '*:*:*': 'deny' },
        },
      },
    });
    const component = render(
      <Provider store={store}>
        <ModuleMitreAttackIntelligence />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
