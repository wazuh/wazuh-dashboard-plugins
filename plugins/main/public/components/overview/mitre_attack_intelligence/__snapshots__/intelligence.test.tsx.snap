// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module Mitre Att&ck intelligence container should render permissions prompt when no has permissions 1`] = `
<div
  class="euiEmptyPrompt"
>
  <svg
    aria-hidden="true"
    class="euiIcon euiIcon--xxLarge euiIcon--subdued euiIcon-isLoading"
    focusable="false"
    height="16"
    role="img"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  />
  <div
    class="euiSpacer euiSpacer--m"
  />
  <h2
    class="euiTitle euiTitle--medium"
  >
    You have no permissions
  </h2>
  <span
    class="euiTextColor euiTextColor--subdued"
  >
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      class="euiText euiText--medium"
    >
      <div>
        This section requires the permission:
        <div>
          <div>
            - 
            <strong>
              mitre:read
            </strong>
             (
            <span
              style="text-decoration:underline"
            >
              *:*:*
            </span>
            )
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`Module Mitre Att&ck intelligence container should render the component if has permissions 1`] = `
<div
  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
  style="margin:0 8px"
>
  <div
    class="euiFlexItem"
  >
    <div
      class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="margin:0"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
          style="background-color:#80808014;margin:0;width:15%;min-width:145px;overflow-x:hidden"
        >
          <div
            style="padding:10px"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--fullWidth"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  aria-label="Search in all resources"
                  class="euiFieldSearch euiFieldSearch--fullWidth"
                  placeholder="Search in all resources"
                  type="search"
                  value=""
                />
                <div
                  class="euiFormControlLayoutIcons"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
          <button
            class="moduleMitreAttackIntelligenceResourceButton moduleMitreAttackIntelligenceResourceButton--selected "
          >
            Groups
          </button>
          <button
            class="moduleMitreAttackIntelligenceResourceButton  "
          >
            Mitigations
          </button>
          <button
            class="moduleMitreAttackIntelligenceResourceButton  "
          >
            Software
          </button>
          <button
            class="moduleMitreAttackIntelligenceResourceButton  "
          >
            Tactics
          </button>
          <button
            class="moduleMitreAttackIntelligenceResourceButton  "
          >
            Techniques
          </button>
        </div>
        <div
          class="euiFlexItem"
          style="max-height:calc(100vh - 255px);overflow-y:auto;overflow-x:hidden"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero wz-flex-basis-auto"
            >
              <h1
                class="euiTitle euiTitle--small"
              >
                Groups 
                <span>
                  (0)
                </span>
              </h1>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
              />
            </div>
          </div>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div>
                <div
                  class="euiSuggestInput"
                >
                  <div
                    class="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiInputPopover--fullWidth"
                    id="popover"
                  >
                    <div
                      class="euiPopover__anchor"
                    >
                      <div>
                        <div
                          class="euiFormControlLayout euiFormControlLayout--fullWidth euiFormControlLayout--group"
                        >
                          <div
                            style="display:flex"
                          />
                          <div
                            class="euiFormControlLayout__childrenWrapper"
                          >
                            <input
                              class="euiFieldText euiFieldText--fullWidth euiFieldText--inGroup"
                              placeholder="Search in Groups"
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--s"
          />
          <div
            class="euiBasicTable"
          >
            <div>
              <div
                class="euiTableHeaderMobile"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                >
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  />
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <div
                      class="euiTableSortMobile"
                    >
                      <div
                        class="euiPopover euiPopover--anchorDownRight"
                      >
                        <div
                          class="euiPopover__anchor"
                        >
                          <button
                            class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                            type="button"
                          >
                            <span
                              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                            >
                              <svg
                                aria-hidden="true"
                                class="euiIcon euiIcon--small euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                focusable="false"
                                height="16"
                                role="img"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              />
                              <span
                                class="euiButtonEmpty__text"
                              >
                                Sorting
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <table
                class="euiTable euiTable--responsive"
                id="htmlId"
                tabindex="-1"
              >
                <caption
                  class="euiScreenReaderOnly euiTableCaption"
                />
                <thead>
                  <tr>
                    <th
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_external_id_0"
                      role="columnheader"
                      scope="col"
                      style="width:12%"
                    >
                      <span
                        class="euiTableCellContent"
                      >
                        <span
                          class="euiTableCellContent__text"
                        >
                          ID
                        </span>
                      </span>
                    </th>
                    <th
                      aria-live="polite"
                      aria-sort="ascending"
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_name_1"
                      role="columnheader"
                      scope="col"
                      style="width:30%"
                    >
                      <button
                        class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                        data-test-subj="tableHeaderSortButton"
                        type="button"
                      >
                        <span
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                          >
                            Name
                          </span>
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--medium euiIcon-isLoading euiTableSortIcon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-live="polite"
                      aria-sort="none"
                      class="euiTableHeaderCell"
                      data-test-subj="tableHeaderCell_description_2"
                      role="columnheader"
                      scope="col"
                    >
                      <button
                        class="euiTableHeaderButton"
                        data-test-subj="tableHeaderSortButton"
                        type="button"
                      >
                        <span
                          class="euiTableCellContent"
                        >
                          <span
                            class="euiTableCellContent__text"
                          >
                            Description
                          </span>
                        </span>
                      </button>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    class="euiTableRow"
                  >
                    <td
                      class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                      colspan="3"
                    >
                      <div
                        class="euiTableCellContent euiTableCellContent--alignCenter"
                      >
                        <span
                          class="euiTableCellContent__text"
                        >
                          No items found
                        </span>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
